TestJobQueueUpgrade panics under PlanResourceChange

The crux of the issue:

```

  "method": "/pulumirpc.ResourceProvider/Diff",
  "request": {
    "id": "terraform-20231205221555324200000003",
    "urn": "urn:pulumi:test::job-queue::aws:batch/computeEnvironment:ComputeEnvironment::sampleComputeEnvironment",
    "olds": {
      "arn": "arn:aws:batch:us-west-2:616138583583:compute-environment/terraform-20231205221555324200000003",
      "computeResources": {
        "ec2Configuration": {
          "imageIdOverride": "",
          "imageType": "ECS_AL2"
        },
```

Note that there is a problem with aws:batch/computeEnvironment:ComputeEnvironment resource where the 5.42.0 provider
wrote an object for ec2Configuration when a list is expected by the new version, as is also reflected in the Pulumi
schema.

So when upgrading a stack with ComputeEnvironment from 5.42.0 to the latest provider there is an implicit type mismatch.

There appears to be a problem where PlanResourceChange makes it a louder failure than the current silent failure.

Both PlanResourceChange=true and PlanResourceChange=false run state upgraders (although the current and desired version
is 0), and ironically the resource does not define any state upgraders at all.

Note that the state cty.Value passed to the state upgraders differs:


```
PlanResourceChange=true

panic: UPGRADING GRPC RAW=cty.TupleVal([]cty.Value{
   cty.ObjectVal(map[string]cty.Value{"allocation_strategy":cty.StringVal(""), "bid_percentage":cty.NumberIntVal(0), "desired_vcpus":cty.NumberIntVal(0),
      "ec2_configuration":cty.ObjectVal(map[string]cty.Value{"image_id_override":cty.StringVal("[]"), "image_type":cty.StringVal("ECS_AL2")}), "ec2_key_pair":cty.StringVal(""), "image_id":cty.StringVal(""), "instance_role":cty.StringVal("arn:aws:iam::616138583583:instance-profile/ecsInstanceRoleInstanceProfile-912858a"), "instance_type":cty.TupleVal([]cty.Value{cty.StringVal("c4.large")}), "max_vcpus":cty.NumberIntVal(16), "min_vcpus":cty.NumberIntVal(0), "security_group_ids":cty.TupleVal([]cty.Value{cty.StringVal("sg-0bb51dcade107992e")}), "spot_iam_fleet_role":cty.StringVal(""), "subnets":cty.TupleVal([]cty.Value{cty.StringVal("subnet-07f3a862bd9c5f1a3")}), "tags":cty.EmptyObjectVal, "type":cty.StringVal("EC2")})}) [recovered]

PlanResourceChange=false

panic: UPGRADING GRPC RAW=cty.ListVal([]cty.Value{
   cty.ObjectVal(map[string]cty.Value{"allocation_strategy":cty.StringVal(""), "bid_percentage":cty.NumberIntVal(0), "desired_vcpus":cty.NumberIntVal(0),
       "ec2_configuration":cty.ListValEmpty(cty.Object(map[string]cty.Type{"image_id_override":cty.String, "image_type":cty.String})), "ec2_key_pair":cty.StringVal(""), "image_id":cty.StringVal(""), "instance_role":cty.StringVal("arn:aws:iam::616138583583:instance-profile/ecsInstanceRoleInstanceProfile-912858a"), "instance_type":cty.SetVal([]cty.Value{cty.StringVal("c4.large")}), "launch_template":cty.ListValEmpty(cty.Object(map[string]cty.Type{"launch_template_id":cty.String, "launch_template_name":cty.String, "version":cty.String})), "max_vcpus":cty.NumberIntVal(16), "min_vcpus":cty.NumberIntVal(0), "placement_group":cty.StringVal(""), "security_group_ids":cty.SetVal([]cty.Value{cty.StringVal("sg-0bb51dcade107992e")}), "spot_iam_fleet_role":cty.StringVal(""), "subnets":cty.SetVal([]cty.Value{cty.StringVal("subnet-07f3a862bd9c5f1a3")}), "tags":cty.MapValEmpty(cty.String), "type":cty.StringVal("EC2")})})

```

Note that PlanResourceChange=false elegantly bypassed the problem by simply ignoring the data it can't read in
ec2_configuration.

However for PlanResoruceChange=true the TupleVal is hiding an underlying failure of recoverCtyValue:

```
recoverCtyValue fails:
panic: FAILED: recoverCtyValue failed: Cannot reconcile map map[image_id_override:[] image_type:ECS_AL2] to cty.List(cty.Object(map[string]cty.Type{"image_id_override":cty.String, "image_type":cty.String})) [recovered]
```

Which makes sense, it failed to read the object into a slot of type list. Then it decided to use approximate methods to
recover the state, which yielded the TupleVal combined with an ec2_configuration that is now an ObjectVal which is not
expected.

Threading this data through resource upgraders trips up JSONMapToStateValue which responds with a very unhelpful error
message: "missing expected [".

There is another issue where the bridge fails to handle the error diagnostics from the state upgraders, and therefore
panics instead of showing this error.


```	github.com/pulumi/pulumi-terraform-bridge/pf v0.36.1-0.20240603141117-4503342dc4c8
	github.com/pulumi/pulumi-terraform-bridge/v3 v3.83.1-0.20240603141117-4503342dc4c8
```

--- a/provider/pkg/batch/compute_environment.go
+++ b/provider/pkg/batch/compute_environment.go
@@ -1,0 +1,50 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package batch
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge/info"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+)
+
+func ComputeEnvironment(token tokens.Type) *info.Resource {
+	return &info.Resource{
+		Tok: token,
+		TransformFromState: func(ctx context.Context, pm resource.PropertyMap) (resource.PropertyMap, error) {
+			// computeResources.ec2Configurations should be a list
+			// older versions of the provider like 5.42.0 used to write it out as an object
+			r := pm.Copy()
+			cr, ok := r["computeResources"]
+			if !ok || !cr.IsObject() {
+				return pm, nil
+			}
+			ec2c, ok := cr.ObjectValue()["ec2Configuration"]
+			if !ok || !ec2c.IsObject() {
+				return pm, nil
+			}
+			replacement := resource.NewArrayProperty([]resource.PropertyValue{ec2c})
+			tfbridge.GetLogger(ctx).Debug(fmt.Sprintf(
+				"batch.ComputeEnvironment is wrapping old computeResources.ec2Configurations state in an array"))
+			cr.ObjectValue()["ec2Configuration"] = replacement
+			r["computeResources"] = cr
+			return r, nil
+		},
+	}
+}

--- a/provider/pkg/batch/compute_environment_test.go
+++ b/provider/pkg/batch/compute_environment_test.go
@@ -1,0 +1,36 @@
+package batch
+
+import (
+	"context"
+	"testing"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/unstable/logging"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/stretchr/testify/require"
+)
+
+func TestComputeEnvironmentTransformFromState(t *testing.T) {
+	ctx := context.Background()
+	ctx = logging.InitLogging(ctx, logging.LogOptions{})
+	pm := resource.PropertyMap{
+		"computeResources": resource.NewObjectProperty(resource.PropertyMap{
+			"ec2Configuration": resource.NewObjectProperty(resource.PropertyMap{
+				"imageIdOverride": resource.NewStringProperty(""),
+				"imageType":       resource.NewStringProperty("ECS_AL2"),
+			}),
+		}),
+	}
+	expect := resource.PropertyMap{
+		"computeResources": resource.NewObjectProperty(resource.PropertyMap{
+			"ec2Configuration": resource.NewArrayProperty([]resource.PropertyValue{
+				resource.NewObjectProperty(resource.PropertyMap{
+					"imageIdOverride": resource.NewStringProperty(""),
+					"imageType":       resource.NewStringProperty("ECS_AL2"),
+				}),
+			}),
+		}),
+	}
+	actual, err := ComputeEnvironment("").TransformFromState(ctx, pm)
+	require.NoError(t, err)
+	require.Equal(t, expect, actual)
+}

--- a/provider/provider_nodejs_test.go
+++ b/provider/provider_nodejs_test.go
@@ -38,6 +38,10 @@ func TestRoute53Upgrade(t *testing.T) {
 func TestJobQueueUpgrade(t *testing.T) {
 	opts := nodeProviderUpgradeOpts()
 	opts.setEnvRegion = false
+	opts.region = "us-west-2" // has to match the snapshot-recorded region
+	opts.extraOpts = []opttest.Option{
+		opttest.Env("PULUMI_ENABLE_PLAN_RESOURCE_CHANGE", "true"),
+	}
 	testProviderUpgrade(t, filepath.Join("test-programs", "job-queue"), opts)
 }
 

--- a/provider/provider_test.go
+++ b/provider/provider_test.go
@@ -31,6 +31,8 @@ type testProviderUpgradeOptions struct {
 	linkNodeSDK     bool
 	installDeps     bool
 	setEnvRegion    bool
+	region          string
+	extraOpts       []opttest.Option
 }
 
 func skipIfShort(t *testing.T) {
@@ -62,10 +64,14 @@ func testProviderUpgrade(t *testing.T, dir string, opts *testProviderUpgradeOpti
 	if opts != nil && opts.linkNodeSDK {
 		options = append(options, opttest.YarnLink("@pulumi/aws"))
 	}
+	options = append(options, opts.extraOpts...)
 	test := pulumitest.NewPulumiTest(t, dir, options...)
 	if opts != nil && opts.setEnvRegion {
 		test.SetConfig("aws:region", "INVALID_REGION")
 		test.SetConfig("aws:envRegion", getEnvRegion(t))
+	}
+	if opts.region != "" {
+		test.SetConfig("aws:region", opts.region)
 	}
 	result := providertest.PreviewProviderUpgrade(
 		test, providerName, baselineVersion, optproviderupgrade.DisableAttach(),

--- a/provider/provider_test.go
+++ b/provider/provider_test.go
@@ -64,7 +64,9 @@ func testProviderUpgrade(t *testing.T, dir string, opts *testProviderUpgradeOpti
 	if opts != nil && opts.linkNodeSDK {
 		options = append(options, opttest.YarnLink("@pulumi/aws"))
 	}
-	options = append(options, opts.extraOpts...)
+	if opts != nil {
+		options = append(options, opts.extraOpts...)
+	}
 	test := pulumitest.NewPulumiTest(t, dir, options...)
 	if opts != nil && opts.setEnvRegion {
 		test.SetConfig("aws:region", "INVALID_REGION")

--- a/provider/provider_test.go
+++ b/provider/provider_test.go
@@ -72,7 +72,7 @@ func testProviderUpgrade(t *testing.T, dir string, opts *testProviderUpgradeOpti
 		test.SetConfig("aws:region", "INVALID_REGION")
 		test.SetConfig("aws:envRegion", getEnvRegion(t))
 	}
-	if opts.region != "" {
+	if opts != nil && opts.region != "" {
 		test.SetConfig("aws:region", opts.region)
 	}
 	result := providertest.PreviewProviderUpgrade(

--- a/provider/resources.go
+++ b/provider/resources.go
@@ -31,6 +31,7 @@ import (
 	awsbase "github.com/hashicorp/aws-sdk-go-base/v2"
 	tfschema "github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/mitchellh/go-homedir"
+	"github.com/pulumi/pulumi-aws/provider/v6/pkg/batch"
 	"github.com/pulumi/pulumi-aws/provider/v6/pkg/rds"
 	"github.com/pulumi/pulumi-aws/provider/v6/pkg/version"
 
@@ -1259,7 +1260,7 @@ func ProviderFromMeta(metaInfo *tfbridge.MetadataInfo) *tfbridge.ProviderInfo {
 				},
 			},
 			// Batch
-			"aws_batch_compute_environment": {Tok: awsResource(batchMod, "ComputeEnvironment")},
+			"aws_batch_compute_environment": batch.ComputeEnvironment(awsResource(batchMod, "ComputeEnvironment")),
 			"aws_batch_job_definition":      {Tok: awsResource(batchMod, "JobDefinition")},
 			"aws_batch_job_queue":           {Tok: awsResource(batchMod, "JobQueue")},
 			"aws_batch_scheduling_policy":   {Tok: awsResource(batchMod, "SchedulingPolicy")},

